### PR TITLE
Move the tracking of unsplittable tablets to metadata table

### LINKED ISSUE #4177: Move tracking of unsplittable tablets from manager memory to metadata table
When a tablet is larger than the split threshold but a split point can not be found (this could happen if the tablet has a very large row with many columns) Accumulo tracks this.  It is tracked to avoid repeatedly wasting resources on trying to find a split point for a tablet that can not split.  In the elasticity branch, tracking of which tablets can not split is done in manager memory [here](https://github.com/apache/accumulo/blob/f6c096bef1c7809b698a9ed9b39d1f625c1e0ee3/server/manager/src/main/java/org/apache/accumulo/manager/split/Splitter.java#L53),

This tracking could be moved to the metadata table by adding a new column to tablet metadata for tablets that are unsplittable.  Tracking this information in the metadata table would reduce memory pressure on the manager and would make running splits in distributed fate possible. Its also nice because the information will survive manager restarts and not possibly cause a flurry of unneeded computation when the manager restarts.

This tracking could be moved to the metadata table by adding a new column to tablet metadata for tablets that are unsplittable.    This column could record the following information.

 * The relevant configuration at the time the tablet was determined to be unsplittable.  For example, the tables split threshold config at the the time could be recorded.  May not need to record the actual config key/values, could possibly record a hash of them.
 * Information about the tablets files at the time the tablet was determined to be unsplittable.  This does not need to be the full set of file URLs.  Could be just the filenames or a hash of the file URIs.  The current code that tracks things in memory tracks a hash of the file URIs to detect a change in the set of files.

The tablet management iterator is used to find tablet that need to split.  This iterator could be modified to use this new column when determining if tablet needs to split.  If the column exists it could check if the table config or tablet files has changed since the column was set.

### PR DESCRIPTION
**Pull request by cshannon**
This adds a new column to store information for tracking unsplittable tablets in the metadata table instead of in memory. This information can be used by the tablet management iterator to know if a tablet needs to split by checking the column and comparing it to the current metadata and avoid unnecessarily trying to split a tablet that can't be split. The data stored includes a hash of the file set and the relevant config related to splits and if this changes then the iterator will try and split again.

The metadata is stored as Json and the configs included include split threshold, table max end row size, and max open files as well as the file set hash. If a split happens on the tablet in the future (because the metadata changed such a the file set and the iterator attempted to split again) the column will be cleaned up by fate in UpdateTablets repo.

For the schema, I wasn't sure of the best approach so I just went ahead and created a new Split column family and unspilttable column qualifier. I thought it might be useful in case we wanted to store more things in the future related to splits in the new column family, however I can easily change the schema if desired or store the new metadata as a column part of an existing column family or just make it a stand alone column family/qualifier like the UserCompactionRequested  or Merged column families.

I marked this as a draft for now as it still needs more tests (only one test was modified so far) to be added and some other clean up work plus as I said I wasn't sure if the schema was the best approach.

This closes #4177

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 1]
```diff
@@ -381,6 +384,11 @@ public boolean getHostingRequested() {
     return onDemandHostingRequested;
   }
 
+  public UnSplittableMetadata getUnSplittable() {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:387]: This comment is not related to this line.  Need to update the toString() method on this class.

**Review by keith-turner — COMMENTED** [review]
Realized there may be cases where the unsplittable marker could just hang around forever, even when its no longer applicable. Something like the following

 * A tablet needs to split, but cannot so the new unsplittable column is written
 * A compaction happens that filters a lot of the tablets data.  The tablet is no longer a candidate for split because it shrunk.

In the case above the unsplittable marker does not get removed.  It does not cause any harm. The only problem with this case is if someone is actively scanning the metadata table looking for problematic tablets then they will find false postives.

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java [hunk 2]
```diff
@@ -69,7 +72,23 @@ public class TabletManagementIterator extends SkippingIterator {
...
+        tableConfig.get(Property.TSERV_TABLET_SPLIT_FINDMIDPOINT_MAXOPEN.getKey()));
+
+    // If current config/files match unsplittable metadata then we can't split
+    if (Optional
+        .ofNullable(tm.getUnSplittable()).filter(um -> um.equals(UnSplittableMetadata
+            .toUnSplittable(splitThreshold, maxEndRowSize, maxFilesToOpen, tm.getFiles())))
+        .isPresent()) {
+      return false;
+    }
+
```
Comment (keith-turner) [inline server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:91]: Maybe this method should always return true if unsplittable column is present and things changed.  This would give the split code in the manager a chance to remove the column if its not needed anymore.  Thinking of something like the following.

```java
if(unsplittable is present AND unsplittableFromTablet != computedUnsplittable){
   // when unsplittable column is out of date always return true, even if tablet is not over
   // split threshold.  This gives code in manager a chance to remove the column.
   return true;
} else if(unsplittable is absent AND tablet is over split threshold){
   return true;
}
```

**Comment by cshannon** [comment]
> In the case above the unsplittable marker does not get removed. It does not cause any harm. The only problem with this case is if someone is actively scanning the metadata table looking for problematic tablets then they will find false postives.

I am thinking to solve this we could update the compaction code to remove the marker when tablets are finished compacting?

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java (hunk 2 above)
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:91]: Ah just saw this comment, this would take away the need from having the compaction code having to deal with the marker so I like this better than having the compaction code deal with cleaning things up as it handles all cases if something changes

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java [hunk 3]
```diff
@@ -69,7 +72,23 @@ public class TabletManagementIterator extends SkippingIterator {
...
+    final long splitThreshold = ConfigurationTypeHelper
+        .getFixedMemoryAsBytes(tableConfig.get(Property.TABLE_SPLIT_THRESHOLD.getKey()));
+    final long maxEndRowSize = ConfigurationTypeHelper
+        .getFixedMemoryAsBytes(tableConfig.get(Property.TABLE_MAX_END_ROW_SIZE.getKey()));
+    final int maxFilesToOpen = (int) ConfigurationTypeHelper.getFixedMemoryAsBytes(
+        tableConfig.get(Property.TSERV_TABLET_SPLIT_FINDMIDPOINT_MAXOPEN.getKey()));
+
+    // If current config/files match unsplittable metadata then we can't split
+    if (Optional
+        .ofNullable(tm.getUnSplittable()).filter(um -> um.equals(UnSplittableMetadata
```
[the remark this answers is not in the document]
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java]: I am switching this to just a normal null check anyways with some refactoring I am doing

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java [hunk 4]
```diff
@@ -75,13 +88,17 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
     if (splits.isEmpty()) {
       log.info("Tablet {} needs to split, but no split points could be found.",
           tabletMetadata.getExtent());
-      // ELASTICITY_TODO record the fact that tablet is un-splittable in metadata table in a new
-      // column. Record the config used to reach this decision and a hash of the file. The tablet
-      // mgmt iterator can inspect this column and only try to split the tablet when something has
-      // changed.
+
+      // TODO: Do we care about conditional mutations here? I don't think it is required because
```
[the remark this answers is not in the document]
Comment (cshannon) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java]: Latest update adds in a conditional mutation, i added the condition to require files be the same as i figure we should fail fast and retry vs having to submit another op later when it detects it's different, i added a comment about it

**Comment by cshannon** [comment]
@keith-turner - This is ready for another review, I realized I had to add a recheck in the FindSplits repo to see if we still are above the split threshold if no split points are found and the marker already exists. This is because the tablet management iterator will now try and split anytime there is a metadata change for an unsplittable tablet and it's possible we no longer need to split (like after a compaction) and we need to clear the marker. Before this change the marker was only cleared after a split was done.

I also added tests for TabletManagement and constraints as well as a couple ITs. There is one IT that will test that the marker is set and then that it is cleared after splits run successfully and a second test that will test that the marker is set and then cleared if we no longer need to split such as the split threshold increasing but no split was done.

**Comment by cshannon** [comment]
One thing to mention is now that FindSplits fate repo has to recheck if a split is needed to clear the marker, it requires having the same logic for checking the threshold in multiple spots so I am wondering if we should consolidate that. For example, now [SplitUtils](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitUtils.java#L215), [FindSplits](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java#L153), and [TabletManagementIterator](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java#L95) all do checks on the split threshold and it seems error prone if we change the logic in the future and one spot is updated and not the others.

Also, all of those spots now need to compute the size by iterating over the files so I was wondering if we should move this into `TabletMetadata` class itself so it could be cached and re-used. Currently I updated the SplitUtils [findSplits](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitUtils.java#L205)() method to now take the estimatedSize as an argument so it's not computed more than once in FindSplits but that seems a bit odd. I was wondering if maybe we should add a getter that can compute the size and cache it (we could use Suppliers.memoize()) and then if any calling code needs to get the size of the files it could just call something like `tabletMetadata.getFileSize()`. The supplier could default to just returning 0 if files are not loaded.

I was also trying to think if we could simplify having to recheck the threshold in the condition of no splits found and the marker exists. SplitUtils findSplits() is already called before that point and already does the threshold check (as noted above) but it doesn't return why it found no splits. Another option is to return a reason from that method for finding no splits (besides the empty set currently) as if the reason returned was the threshold wasn't met then we could avoid having to recheck.

The only other thing was I was trying to think of if there are more test cases we should write.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+import com.google.gson.Gson;
+
+public class UnSplittableMetadata {
+
+  private static final Gson GSON = ByteArrayToBase64TypeAdapter.createBase64Gson();
+
+  private final long splitThreshold;
+  private final long maxEndRowSize;
+  private final int maxFilesToOpen;
+  private final HashCode filesHash;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:42]: Thinking it may simplify the code in this class if all we store is a hash.  Something like the following.  Would remove the need for the json code.
```suggestion
  // stores a base64 encoded murmur 128 bit hash splitThreshold,maxEndRowSize,maxFilesToOpen, set of files.
  private final String hashOfSplitParameters;
```

**Review by keith-turner — COMMENTED** [review]
Still reviewing changes, posting comment made so far.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java (hunk 5 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:42]: This would definitely simply things but can you think of a reason we may need to read the original values? (maybe logging?) This change is probably fine as I don't think we need to query those stored values individually again but figure I'd double check before changing it. We could also log the values that are used to create the hash (maybe as trace) so we could access it there.

Comment (keith-turner) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:42]: > can you think of a reason we may need to read the original values?

The only case I can think of currently is debugging a problem in the code and I think for that case your suggestion to log covers that.

> We could also log the values that are used to create the hash (maybe as trace) so we could access it there.

I think logging the inputs to the hash and the resulting hash would be very useful.  Then if there is a problem can search the logs for the hash.

**Comment by keith-turner** [comment]
>  I was wondering if maybe we should add a getter that can compute the size and cache it (we could use Suppliers.memoize()) and then if any calling code needs to get the size of the files it could just call something like tabletMetadata.getFileSize(). The supplier could default to just returning 0 if files are not loaded.

That sounds nice, could  throw an exception instead of return zero if files were not fetched.

**Comment by keith-turner** [comment]
> One thing to mention is now that FindSplits fate repo has to recheck if a split is needed to clear the marker, it requires having the same logic for checking the threshold in multiple spots so I am wondering if we should consolidate that. For example, now [SplitUtils](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/split/SplitUtils.java#L215), [FindSplits](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java#L153), and [TabletManagementIterator](https://github.com/apache/accumulo/blob/cb0f9959de1d223d6d76fbb13ffed7cc5e144867/server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java#L95) all do checks on the split threshold and it seems error prone if we change the logic in the future and one spot is updated and not the others.

Consolidating that decision logic to a single class would be really nice and less error prone.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java [hunk 6]
```diff
@@ -66,22 +91,68 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
           tabletMetadata.getLogs().size());
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:91]: This is preexisting problem, seems like it should return null here.

```suggestion
          tabletMetadata.getLogs().size());
          return null;
```

File: test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java [hunk 7]
```diff
@@ -242,6 +257,138 @@ public void automaticSplitLater() throws Exception {
...
+
+      final int numOfMutations = 20;
+      try (BatchWriter batchWriter = client.createBatchWriter(tableName)) {
+        // Make the last place in the key different for every entry added to the table
+        for (int i = 0; i < numOfMutations; i++) {
+          data[data.length - 1] = (byte) i;
+          Mutation m = new Mutation(data);
+          m.put("cf", "cq", "value");
+          batchWriter.addMutation(m);
+        }
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java:367]: There is another condition that can cause a tablet to be unsplittable and that is if every key has the same row.  This test could exercise that condition while `testUnsplittableColumn()` excercises the other condition that causes an unsplittable tablet.  Attempted to modify the code to create this other condition below. This way we get a check that both situations cause the marker to appear.


```suggestion
        for (int i = 0; i < numOfMutations; i++) {
          Mutation m = new Mutation(data);
          m.put("cf", "cq"+i, "value");
          batchWriter.addMutation(m);
        }
```

**Review by keith-turner — COMMENTED** [review]
Finished review everything.

**Comment by cshannon** [comment]
@keith-turner - looks like your new test in #4323 caused this PR to break...which is good news as that was the intent. I will update this PR to add the column to the test

```
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   UpdateTabletsTest.checkColumns:100 The split code does not know how to handle UNSPLITTABLE ==> expected: <true> but was: <false>
[ERROR]   UpdateTabletsTest.testManyColumns:363 
  Unexpected method call TabletMetadata.getUnSplittable():
[INFO] 
[ERROR] Tests run: 56, Failures: 2, Errors: 0, Skipped: 0
```

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java [hunk 8]
```diff
@@ -69,7 +71,25 @@ public class TabletManagementIterator extends SkippingIterator {
   private TabletBalancer balancer;
 
   private static boolean shouldReturnDueToSplit(final TabletMetadata tm,
-      final long splitThreshold) {
+      final Configuration tableConfig) {
```
[the remark this answers is not in the document]
Comment (cshannon) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:74]: The problem with trying to put it into SplitUtils is that class is part of the manager module and TabletManagementIterator is part of base module so it's not visible as base doesn't have a dependency on manager (it's reversed)

Maybe SplitUtils should be moved to base?

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java (hunk 5 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java]: Changed in in 6c3358b5783b19ec6d73c7f7a121d0b0493c4333

File: core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java [hunk 9]
```diff
@@ -339,7 +344,47 @@ public void testCompactionRequestedColumn() {
...
+    StoredTabletFile sf3 = StoredTabletFile.of(new Path("hdfs://nn1/acc/tables/1/t-0001/sf3.rf"));
+    var unsplittableMeta =
+        UnSplittableMetadata.toUnSplittable(100, 110, 120, Set.of(sf1, sf2, sf3));
+
+    // Test with files
+    Mutation mutation = TabletColumnFamily.createPrevRowMutation(extent);
+    SplitColumnFamily.UNSPLITTABLE_COLUMN.put(mutation, new Value(unsplittableMeta.toJson()));
+    TabletMetadata tm = TabletMetadata.convertRow(toRowMap(mutation).entrySet().iterator(),
+        EnumSet.of(UNSPLITTABLE), true, false);
+    assertEquals(unsplittableMeta, tm.getUnSplittable());
```
[the remark this answers is not in the document]
Comment (cshannon) [reply core/src/test/java/org/apache/accumulo/core/metadata/schema/TabletMetadataTest.java]: Test was updated in 6c3358b5783b19ec6d73c7f7a121d0b0493c4333

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java [hunk 10]
```diff
@@ -66,22 +91,68 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
...
+              .requireSame(tabletMetadata, FILES).setUnSplittable(unSplittableMeta);
+          mutator.submit(tm -> unSplittableMeta.equals(tm.getUnSplittable()));
+        } else {
+          // We no longer need to split so we can clear the marker.
+          log.info("Tablet {} no longer needs to split, deleting unsplittable marker.",
+              tabletMetadata.getExtent());
+          var mutator = tabletsMutator.mutateTablet(extent).requireAbsentOperation()
+              .requireSame(tabletMetadata, FILES).deleteUnSplittable();
+          mutator.submit(tm -> tm.getUnSplittable() == null);
+        }
```
[the remark this answers is not in the document]
Comment (cshannon) [reply server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:141]: Good catch, this avoids having try try and clear the marker if it is already null and never set

File: server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply server/base/src/main/java/org/apache/accumulo/server/manager/state/TabletManagementIterator.java:73]: > Maybe SplitUtils should be moved to base?

Moving it to server base SGTM

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java [hunk 11]
```diff
@@ -316,6 +322,11 @@ public Map<StoredTabletFile,DataFileValue> getFilesMap() {
     return files;
   }
 
+  public long getFileSize() {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:325]:
```suggestion
   /**
    * @return the sum of the tablets files sizes
    */
  public long getFileSize() {
```


Would be nice to add a unit test for this new method.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  public String toString() {
+    return toBase64();
+  }
+
+  public String toBase64() {
+    return Base64.getEncoder().encodeToString(hashOfSplitParameters.asBytes());
+  }
+
+  @SuppressWarnings("UnstableApiUsage")
+  private static HashCode caclulateSplitParamsHash(long splitThreshold, long maxEndRowSize,
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:73]: Not really sure if implementing this comment is worthwhile, but also can not think of any harm implementing it may cause.

If a tablet is merged and this column is present, then merge should remove it.  To make the hash itself more strict, wondering if we should add the extent into the hash calculation.  That way if a merge happened and the column was not removed, then the hash would still differ if the extent is the only thing that is not the same between the pre and post merge tablet.  Maybe this overkill though.


```suggestion
  private static HashCode caclulateSplitParamsHash(KeyExtent extent, long splitThreshold, long maxEndRowSize,
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+    files.stream().map(StoredTabletFile::getNormalizedPathStr).sorted()
+        .forEach(path -> hasher.putString(path, UTF_8));
+    return hasher.hash();
+  }
+
+  public static UnSplittableMetadata toUnSplittable(String base64HashOfSplitParameters) {
+    return toUnSplittable(Base64.getDecoder().decode(base64HashOfSplitParameters));
+  }
+
+  public static UnSplittableMetadata toUnSplittable(byte[] hashOfSplitParameters) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:93]: I was looking at the MetadataConstraint class and wondering why it did not call this method because that code has a byte array.  Looking into it I think this method does not handle base64.  This could be private because nothing external calls it and its different expecations for input is very subtle.  Could inline it instead of making it private.

```suggestion
  private static UnSplittableMetadata toUnSplittable(byte[] hashOfSplitParameters) {
```

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java (hunk 11 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/TabletMetadata.java:325]: Yeah, I still need to do some more testing updates for the ITs as well so i'll add it at the same time

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java (hunk 12 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:73]: Seems fine to add the extent as well, there is no harm since it is supposed to be per tablet anyways.

File: test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java (hunk 7 above)
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java]: @keith-turner - the latest commit efe30445ad1e69151bdac4e93ac57f05d05d8e4e updates this test by writing the same row (that is not too large to split) as suggested and writes enough data so we end up splitting the default tablet and then we end up with a tablet that is unsplittable due to this condition. The test then still bumps the threshold and verifies all the tablets get cleared of the marker and no more splits happen. See what you think..we could also keep the old test and add this as a 3rd test too.

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+    Preconditions.checkArgument(maxEndRowSize > 0, "maxEndRowSize must be greater than 0");
+    Preconditions.checkArgument(maxFilesToOpen > 0, "maxFilesToOpen must be greater than 0");
+
+    // Use static call to murmur3_128() so the seed is always the same
+    // Hashing.goodFastHash will seed with the current time, and we need the seed to be
+    // the same across restarts and instances
+    var hasher = Hashing.murmur3_128().newHasher();
+    hasher.putBytes(serializeKeyExtent(keyExtent)).putLong(splitThreshold).putLong(maxEndRowSize)
+        .putInt(maxFilesToOpen);
+    files.stream().map(StoredTabletFile::getNormalizedPathStr).sorted()
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:90]: Would the following also include the range associated with the file?  If so that may be good to include.

```suggestion
    files.stream().map(StoredTabletFile::getMetadata).sorted()
```

If this is changed, then could update the unit test to have some files with ranges.

File: server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java [hunk 15]
```diff
@@ -73,15 +95,59 @@ public Repo<Manager> call(FateId fateId, Manager manager) throws Exception {
...
+        } else if (tabletMetadata.getUnSplittable() != null) {
+          log.info("Tablet {} no longer needs to split, deleting unsplittable marker.",
+              tabletMetadata.getExtent());
+          var mutator = tabletsMutator.mutateTablet(extent).requireAbsentOperation()
+              .requireSame(tabletMetadata, FILES).deleteUnSplittable();
+          mutator.submit(tm -> tm.getUnSplittable() == null);
+          // Case 3: The table config and/or set of files changed since the tablet mgmt iterator
+          // examined this tablet.
+        } else {
+          log.info("Tablet {} no longer needs to split, ignoring it.", tabletMetadata.getExtent());
```
Comment (keith-turner) [inline server/manager/src/main/java/org/apache/accumulo/manager/tableOps/split/FindSplits.java:143]: I would lower the log level on this, unless there is a bug with splitting its not useful.


```suggestion
          log.debug("Tablet {} no longer needs to split, ignoring it.", tabletMetadata.getExtent());
```

File: server/manager/src/test/java/org/apache/accumulo/manager/tableOps/split/UpdateTabletsTest.java [hunk 16]
```diff
@@ -261,6 +261,7 @@ public void testManyColumns() throws Exception {
     EasyMock.expect(tabletMeta.getHostingRequested()).andReturn(true).atLeastOnce();
     EasyMock.expect(tabletMeta.getSuspend()).andReturn(suspendingTServer).atLeastOnce();
     EasyMock.expect(tabletMeta.getLast()).andReturn(lastLocation).atLeastOnce();
+    EasyMock.expect(tabletMeta.getUnSplittable()).andReturn(null).atLeastOnce();
```
Comment (keith-turner) [inline server/manager/src/test/java/org/apache/accumulo/manager/tableOps/split/UpdateTabletsTest.java:264]: Would be good to test the case where its present to ensure its deleted.

```suggestion
    UnSplittableMetadata usm = UnSplittableMetadata.toUnSplittable(origExtent, 1000, 1001, 1002, tabletFiles.keySet());
    EasyMock.expect(tabletMeta.getUnSplittable()).andReturn(usm).atLeastOnce();
```

Then later in the code where it checks what methods were called on the mutator could do the following.

```java
EasyMock.expect(tablet3Mutator.deleteUnSplittable()).andReturn(tablet3Mutator);
```

File: test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java [hunk 17]
```diff
@@ -242,6 +257,148 @@ public void automaticSplitLater() throws Exception {
...
+          Wait.MAX_WAIT_MILLIS, 100);
+
+      // Verify that the unsplittable column is read correctly
+      TabletMetadata tm =
+          getServerContext().getAmple().readTablet(new KeyExtent(tableId, null, null));
+      assertEquals(tm.getUnSplittable(), SplitUtils.toUnSplittable(getServerContext(), tm));
+
+      // Make sure no splits occurred in the table
+      assertTrue(client.tableOperations().listSplits(tableName).isEmpty());
+
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java:311]: Would be nice to squeeze the following in here.

 * record the current unsplittable value
 * set Property.TABLE_MAX_END_ROW_SIZE to 101
 * Wait for the unsplittable value to change (assuming the tablet still needs to split but can not)
 * set Property.TABLE_MAX_END_ROW_SIZE to 500 and continue with rest of test

File: core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java (hunk 14 above)
Comment (cshannon) [reply core/src/main/java/org/apache/accumulo/core/metadata/schema/UnSplittableMetadata.java:90]: This is a good catch, we definitely should include the full json and metadata so we have the path and any ranges

File: test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java (hunk 17 above)
Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java:311]: That's a good idea to test that, i will add that.

Comment (cshannon) [reply test/src/main/java/org/apache/accumulo/test/LargeSplitRowIT.java:312]: I added this to the test in 12ee73ebc5ecc244681fc49e5528045f330ba995

**Review by keith-turner — APPROVED** [review]
